drivers/bluetooth/hci/h4.c

```diff
@@ -159,15 +161,82 @@ static void reset_rx(void)
 
 static struct net_buf *get_rx(int timeout)
 {
+	struct net_buf *buf;
+
 	BT_DBG("type 0x%02x, evt 0x%02x", rx.type, rx.evt.evt);
 
 	if (rx.type == H4_EVT) {
-		return bt_buf_get_evt(rx.evt.evt, timeout);
+		buf = bt_buf_get_evt(rx.evt.evt, timeout);
+	} else {
+		buf = bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	}
+
+	if (buf) {
+		if (rx.discardable) {
+			buf->flags |= BUF_DISCARDABLE;
+		} else {
+			buf->flags &= ~BUF_DISCARDABLE;
+		}
 	}
 
-	return bt_buf_get_rx(BT_BUF_ACL_IN, timeout);
+	return buf;
 }
 
+#if (CONFIG_BT_H4_DISCARD_RX_WAIT >= 0)
+static struct net_buf *get_rx_blocking(void)
+{
+	sys_slist_t list = SYS_SLIST_STATIC_INIT(&list);
+	bool discarded = false;
+	struct net_buf *buf;
+
+	/* If we have ACL flow control enabled then ACL buffers come from a
+	 * dedicated pool, and we cannot benefit from trying to discard
+	 * events from the RX queue.
+	 */
+	if (IS_ENABLED(CONFIG_BT_HCI_ACL_FLOW_CONTROL) && rx.type == H4_ACL) {
+		return get_rx(K_FOREVER);
```

Comment:
I thought the nature of such patches is to make sure that buffer-getting functions are *never* called with K_FOREVER.

Reply:
The scenario here is that we're in the RX thread with the UART RX interrupt disabled, i.e. there's nothing else that can be done than to sit and wait for a buffer. With ACL flow control this should be guaranteed to succeed since otherwise the controller is sending us an ACL packet when we haven't given it permission (credits) to do so. Another option would be K_NO_WAIT + ASSERT.

Reply:
>  i.e. there's nothing else that can be done than to sit and wait for a buffer.

There's always something to do, e.g. to report a warning after some period of time, that this extended wait happens.

That said, that's just a random comment from someone not familiar with the BT subsys. (In IP networking, I believe we decided that we don't want any K_FOREVER's, and it was even implemented IIRC).

Reply:
For debugging purposes something that might be useful, but it should be made for the entire stack. This actually gave me a nice debug feature idea for the entire kernel: have a Kconfig option, which when enabled causes any threads using K_FOREVER sleep to wake up periodically after a certain amount of time (e.g. every minute) and announce to the system log that they're still sleeping. The kernel APIs wouldn't return, rather just go back to waiting, i.e. the semantics of the APIs being guaranteed to return non-NULL with K_FOREVER wouldn't change.

Reply:
> For debugging purposes something that might be useful, but it should be made for the entire stack. This actually gave me a nice debug feature idea for the entire kernel: have a Kconfig option, which when enabled causes any threads using K_FOREVER sleep to wake up periodically after a certain amount of time (e.g. every minute) and announce to the system log that they're still sleeping. The kernel APIs wouldn't return, rather just go back to waiting, i.e. the semantics of the APIs being guaranteed to return non-NULL with K_FOREVER wouldn't change.

Interrestingm I was thinking something similar sometime ago, though if we just wake up the threads just to make them wait again that perhaps would change the order in the waiters list so we it might have to be done without touching the waiters list.